### PR TITLE
[FLINK-23363][table-code-splitter] Java code splitter now supports functions with return statements

### PR DESCRIPTION
**Pull request by tsreaper**
## What is the purpose of the change

Currently Java code splitter only supports functions with no return value and no return statements. However generated classes like `GeneratedProjection` and `GeneratedRowComparator` may contain long functions with return statements.

This PR supports the splitting of these functions.

## Brief change log

 - Java code splitter now supports functions with return statements

## Verifying this change

This change added tests and can be verified by running the added tests.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): **yes**
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? not applicable

### CODE REVIEW

File: flink-table/flink-table-code-splitter/src/main/java/org/apache/flink/table/codesplit/AddBoolBeforeReturnRewriter.java [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+ *         return;
+ *     }
+ * }
+ * </code></pre>
+ *
+ * <p><i>After</i>
+ *
+ * <pre><code>
+ * public class Example {
+ *     boolean codeSplitHasReturned$0 = false;
```
Comment (JingsongLi) [inline flink-table/flink-table-code-splitter/src/main/java/org/apache/flink/table/codesplit/AddBoolBeforeReturnRewriter.java:61]: `myFunHasReturned$0` is better?

File: flink-table/flink-table-code-splitter/src/main/java/org/apache/flink/table/codesplit/DeclarationRewriter.java [hunk 2]
```diff
@@ -117,6 +119,17 @@ public Void visitClassBody(JavaParser.ClassBodyContext ctx) {
             return ret;
         }
 
+        @Override
+        public Void visitFieldDeclaration(JavaParser.FieldDeclarationContext ctx) {
```
Comment (JingsongLi) [inline flink-table/flink-table-code-splitter/src/main/java/org/apache/flink/table/codesplit/DeclarationRewriter.java:123]: Is this a separate bug?

File: flink-table/flink-table-code-splitter/src/main/java/org/apache/flink/table/codesplit/AddBoolBeforeReturnRewriter.java [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+    public List<Map<String, String>> getBoolVarNames() {
+        return boolVarNames;
+    }
+
+    private class OuterVisitor extends JavaParserBaseVisitor<Void> {
+
+        @Override
+        public Void visitClassBody(JavaParser.ClassBodyContext ctx) {
+            int classCount = boolVarNames.size();
+            boolVarNames.add(new HashMap<>());
```
Comment (JingsongLi) [inline flink-table/flink-table-code-splitter/src/main/java/org/apache/flink/table/codesplit/AddBoolBeforeReturnRewriter.java:116]:
```
Map classBoolNames = new HashMap();
boolVarNames.add(classBoolNames);
```
We dont need have a confused `classCount` field.

File: flink-table/flink-table-code-splitter/src/main/java/org/apache/flink/table/codesplit/FunctionSplitter.java [hunk 4]
```diff
@@ -50,26 +59,44 @@
  *
  * <pre><code>
  * public class Example {
+ *     boolean codeSplitHasReturned$0 = false;
```
Comment (JingsongLi) [inline flink-table/flink-table-code-splitter/src/main/java/org/apache/flink/table/codesplit/FunctionSplitter.java:62]: Document is wrong... This should not in here.

File: flink-table/flink-table-code-splitter/src/main/java/org/apache/flink/table/codesplit/AddBoolBeforeReturnRewriter.java [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+        @Override
+        public Void visitLambdaBody(JavaParser.LambdaBodyContext ctx) {
+            // skip lambda
+            return null;
+        }
+
+        @Override
+        public Void visitStatement(JavaParser.StatementContext ctx) {
+            if (ctx.RETURN() != null) {
+                String newReturnStatement = String.format("{ %s = true; return; }", boolVarName);
```
Comment (JingsongLi) [inline flink-table/flink-table-code-splitter/src/main/java/org/apache/flink/table/codesplit/AddBoolBeforeReturnRewriter.java:176]: Can you add a comment to explain why we need `{ }`.

**Review by JingsongLi — CHANGES_REQUESTED** [review]
Thanks @tsreaper , this really looks nice! Left some comments.

File: flink-table/flink-table-code-splitter/src/main/java/org/apache/flink/table/codesplit/DeclarationRewriter.java (hunk 2 above)
Comment (tsreaper) [reply flink-table/flink-table-code-splitter/src/main/java/org/apache/flink/table/codesplit/DeclarationRewriter.java:123]: Yes it is. I'll clean up the commit.

**Review by JingsongLi — APPROVED** [review]
Thanks for the update! Looks good to me.

**Pull request closed without merging**